bwmf/kldiv_loss.go

```diff
@@ -59,19 +59,21 @@ func (l *KLDivLoss) Evaluate(H op.Parameter, gradient op.Parameter) float32 {
 	grad_data := gradient.Data()
 	w_data := l.W.data.Val
 
+	s_w_data := make([]float32, K)
+
 	for i := uint32(0); i < M; i++ {
-		for j := uint32(0); j < N; j++ {
-			wh := float32(0.0)
-			for k := uint32(0); k < K; k += 1 {
-				wh += w_data[i*K+k] * H_data[j*K+k]
-			}
+		for k := uint32(0); k < K; k += 1 {
+			s_w_data[k] += w_data[i*K+k]
+		}
+	}
 
-			// v is zero
-			value += wh
-			for k := uint32(0); k < K; k += 1 {
-				grad_data[j*K+k] += w_data[i*K+k]
-			}
+	for j := uint32(0); j < N; j++ {
+		wh := float32(0)
+		for k := uint32(0); k < K; k += 1 {
+			grad_data[j*K+k] += s_w_data[k]
+			wh += s_w_data[k] * H_data[j*K+k]
 		}
+		value += wh
 	}
 
```

Comment:
This turns the original M*N*K ops into (M+N)*K, by converting `Sum_i Sum_j Sum_k W_ik*Hjk` into `Sum_j Sum_k (H_jk*(Sum_i W_ik))`. 

I suggest that we add the description into the comments.